packages/core/src/typegenTypes.ts

```diff
@@ -0,0 +1,260 @@
+import {
+  ActionObject,
+  ActionFunction,
+  Cast,
+  ConditionPredicate,
+  BaseActionObject,
+  DelayConfig,
+  EventObject,
+  InvokeCreator,
+  IndexByType,
+  IsNever,
+  MachineOptions,
+  Values
+} from './types';
+
+export interface TypegenDisabled {
+  '@@xstate/typegen': false;
+}
+export interface TypegenEnabled {
+  '@@xstate/typegen': true;
+}
+export interface TypegenMeta extends TypegenEnabled {
+  matchesStates: string;
+  tags: string;
+  missingImplementations: {
+    actions: string;
+    delays: string;
+    guards: string;
+    services: string;
+  };
+  eventsCausingActions: Record<string, string>;
+  eventsCausingDelays: Record<string, string>;
+  eventsCausingGuards: Record<string, string>;
+  eventsCausingServices: Record<string, string>;
+}
+
+export interface ResolvedTypegenMeta extends TypegenMeta {
+  indexedActions: Record<string, BaseActionObject>;
+  indexedEvents: Record<string, EventObject>;
+}
+
+export type TypegenConstraint = TypegenEnabled | TypegenDisabled;
+
+// if combined union of all missing implementation types is never then everything has been provided
```

Comment:
This is a neat trick that allows us to compare again **fewer** strings than more (since the check is inversed). In a way we don't have to match against any strings - we just create a union and check if the result is `never`.